### DISCUSSION TITLE
Developer discoverability

### DISCUSSION THREAD

**Discussion by joemasilotti** [joemasilotti/railsdevs.com#639] [Ideas]
With almost 700 (!) developers on [RailsDevs](https://railsdevs.com/) it's becoming harder for individual folks to stand out after they fall off the homepage. Here are some ideas to resurface developers to potential employers.

## Recently updated

1. Sort by most recently updated
2. A “Recently updated” badge

Both of these can help surface developers who recently tweaked their profile. This is great if someone joined the site a while back but is now looking for something new. Or revamped their profile to better position themselves.

The downside is that it might encourage folks to make frequent, tiny changes to ensure they stay at the top of the list.

## Spotlight/feature developers

Currently, admins can feature a developer for one week. They appear at the top of search results when no additional criteria is selected.

This works but requires too much manual intervention to be effective in its current form. Something more automated and less prone to potential bias is required.

## Prompts + responses

Admins could define a set list of Rails-related questions for developers to answer and display on their profile. This could help businesses learn more about the candidate before reaching out.

The downside is that it would be hard to surface these in search results – it only really helps while you are on a developer’s profile page.

Alternatively, a digest of answers to a single question could be sent out at a regular cadence. This has the added benefit of offering paid subscribers another benefit with their subscription or keeping RailsDevs top-of-mind for folks who haven’t paid, yet.

## Highlight source contributors

Since [RailsDevs is open source](https://github.com/joemasilotti/railsdevs.com) it could make sense to highlight contributors. A badge on search results would highlight folks quite distinctly from others in the list. The badge could link to their commits on GitHub, too.

It could also be extended to “recent contributors” (last X months) to discourage a single contribution and then never touching the codebase again.

The downside for this idea is it would require OAuth with GitHub. While not a huge lift, it would be a decent chunk of work in implementation, UX, and from the developers themselves.

## Random or rotated search results

Randomly ordering the search results could be a fun way to surface different folks without any additional work on the developer’s end. The downside is that revisiting a saved search result (via URL params) and the order being different could be unsettling.

## Expertise or interest badges

Badges set by the developer could help differentiate them from the rest of the list.

These could be limited to 3 or 5 badges/tags from a long list of available options. Limiting the selection would force someone to pick their most valuable attributes, not just something they’ve done once or twice.

The downside if curating and maintaining this list. Even if everyone is surveyed it would require constant maintenance as new technologies or best practices emerge. Not to mention the UX implementation probably needs aliases and other affordances to make it easy to work with.

## High response rate badges

Building on the badge idea, developers could be tagged when they respond to 100% of businesses on the platform. This encourages developers to let businesses know when a company isn’t a good fit and helps businesses find folks who are still active on the site.

## Endorsements from other developers

Endorsements could fill that “referral” gap in the space in a nice way. A developer could endorse another for time they spent working together on a project or at a previous company. It could start small with a thumbs up or boost and morph into more freeform text.

I like this because we could show the endorser’s avatar (or multiple avatars) in the search results making it easy to see folks who have been endorsed by others on the site. But I worry that endorsements might be meaningless if the business doesn’t recognize any of the names.

Another risk is that even the MVP of this idea could require a lot of work before proving if it is viable (and helpful to businesses).

## Pinned projects

Similar to a [GitHub profile’s pinned projects](https://github.com/joemasilotti/), this can help developers share open source projects they contribute to or maintain. It could be a good way to direct someone to specific work instead of sending them to their GitHub profile blindly.

My hesitation with this is that it doesn’t translate well to search results and if it’s worth the investment over having someone pin the same projects on GitHub.

**Comment by joemasilotti** [comment]
The first feature of this idea has been documented in #640, adding a "Recently active" badge to developer profile cards in search results.

**Reply by helioborges** [reply]
This idea is great, thank you for #641 and #649. 

Now, maybe there could be a way to sort results from the last time the user received the badge, with an internal cooldown to prevent the behavior described on your post.

First there could be an option to search only users with the recently_active badge.

Second, when updating the user profile, check if a column (something like last_time_recently_active) is empty or older than a previously decided cooldown (something like RECENTLY_ACTIVE_COOLDOWN). 

If true, it writes the current time in it, so we can now have a new option to sort the results by 'last_time_recently_active desc'

It won't write it over until the cooldown is older, so it will only change if the user updates its profile later (let's say beyond RECENTLY_ACTIVE_LENGTH that it's currently 1 week, or maybe 2x, 3x, or 4x that). 

Just an idea. Thank you joe for everything :)

**Comment by joemasilotti** [comment]
Second feature has been documented in #641, showing the (now two!) badged on the developer profile page.

**Comment by KarlHeitmann** [comment]
I want to share a wild idea here.

One of the cool things of internet, is that you can integrate different web pages via API calls.

Maybe you can use the information provided by the dev on the section `Online presence` and take the github and stackoverflow profile, made some API calls to github and stackoverflow, and retrieve the number of questions asked, the number of responses, the number of PR, etc. And integrate them in the developer profile. The developer could click a link on his profile to launch the job and update his github and stackoverflow data. This way, only an active developer on the site that wants to get noticed will force himself to be more active on github and stackoverflow in order to click the link and update his integration profile. You can put on some links in `/developers` to categorize developers: show developers ranked by number of PR, number of issues submitted, number of reviews done, etc. It is a different profile someone who has a great number of PR, compared to someone with a lot of reviews. Someone who has a great number of reviews can de a good candidate for a tech lead position, someone who has a lot of commits can be a good candidate to solve specific things.

**Reply by joemasilotti** [reply]
Hey @KarlHeitmann, thanks for sharing! This part resonates the most with me:

> Someone who has a great number of reviews can de a good candidate for a tech lead position, someone who has a lot of commits can be a good candidate to solve specific things.

It makes me wonder if folks might want to "tag" themselves a specific type of developer. Folks who contribute code are looking for different roles than management – same goes for high-level architecture positions.

Outside of that, integrating a "score" of sorts that takes Stack Overflow karma along with recent GitHub activity could be a fun way to gamify some of this! There's a lot of risk involved with folks inflating these numbers, but it could be an interesting experiment. I'd also want to make sure that this is actually valuable information – not everyone even has a SO profile.

**Reply by KarlHeitmann** [reply]
Anytime @joemasilotti , pleased to be of assistance.

That's a great idea. I think it's a good improvement to categorize the developers on the platform, so companies can search the profiles they are looking for.

It's a real concern there is people inflating their numbers. But there are some things that are easy to inflate, and some things that take time to inflate. For example the number of PR or reviews are difficult to inflate, because you need another account to accept PR, or ask you for a review. If someone has multiple accounts to do that, he can spend a lot of time to inflate these numbers, but if you weigh multiple third parties score and karma (eg: SO karma, number of PR in github, codewars kata solved, etc) into a single "external score field" on railsdevs, it will be very difficult to inflate the overall score in your platform.

There is something very cool about this project that I want to address. I work in education, and a problem I've seen with my students, is that even though they can do their homework, they can take courses on udemy, etc. This is all individual work -that is **very necessary to have**- that can improve their skillsets, but that's all individual work done in a "laboratory" context: usually they are typical problems (eg: create a twitter clone, create a chart to buy things, etc), and they lack the nuances you usually found on real world problems you need to solve in a real world company. 

In your project you create easy tasks on the `issues` tab, and they are open so anybody can pick up an issue and try to solve it. To solve these problems you need to read code of other people you don't know, you must use strategies to found on the source code the things you need to change based on the screenshots you provide, you must work with constraints, you must stick to the style guide on this project, you must run the tests, etc.

This brings a good opportunity for developers who want to build their skillset using real world problems to supplement the skills they already have built using their individual study. This is very cool, you may even build a positive feedback loop, where people take an issue, submit a PR, get the review, finish the issue, grows his knowledge in the platform, take a little more complex next issue, and loop over again.

To take the score of third party companies (github, SO, codewars, etc) it's a complex task. No doubt. But maybe it can be a "spin-off" project? and put there a list of issues so people can contribute and build their skillset in tasks you don't find in other academies or recorded videos. I can share the project with people on the academy I work as part time professor. To be honest, I don't know if it's worth to make a "spin-off" project, or the task is more easy than what I imagine. I'm just throwing ideas.

**Reply by joemasilotti** [reply]
> To take the score of third party companies (github, SO, codewars, etc) it's a complex task. No doubt. But maybe it can be a "spin-off" project? and put there a list of issues so people can contribute and build their skillset in tasks you don't find in other academies or recorded videos. I can share the project with people on the academy I work as part time professor. To be honest, I don't know if it's worth to make a "spin-off" project, or the task is more easy than what I imagine. I'm just throwing ideas.

A "code score" project could be a fun way to scope this feature while also enabling other folks to use the code in their projects, too. I like that idea. I'm not sure if I have the time to manage another repo, though.

I think a lot of the value you described above comes from the fact that RailsDevs is already live in production and people are using it every day. Starting a new project could end up being like the "laboratory" context issue you bring up – it might be more contrived than a real-world problem.

What do you think?

**Comment by joemasilotti** [comment]
In chatting with David of [jsdevs](https://jsdevs.co) (a fork of RailsDevs) an interesting idea came up.

Developers get an email 30 days after the last update to their profile. This asks if it is still up to date, sees if they got hired, and asks to reply if all looks good.

While great in theory I bet a bunch of folks never see this email. Which means that these now "stale" profiles are still sitting around collecting dust. **Marking these profiles as "not interested" could go a long way in surfacing the _other_ folks on the site that are active and still looking!**

Two ideas on how we can address this:

1. In the 30-day checkup email let the developer know their profile will be marked as unavailable if they don't update it in 7 days.
2. If a developer responds to a message then consider their profile "active" for another 30 days.

**Reply by ghost** [reply]
sorry for duplicating this here - https://github.com/joemasilotti/railsdevs.com/discussions/721. I didn't read this thread earlier. Anyways.

Random displaying could be the solution which is automated. And,most of the hiring managers would bookmark a developer(s) profile rather than the listing page. And as anyways with new developer profiles always showing up on top bookmarking the list doesn't make sense at first place.

**Comment by matteeyah** [comment]
This is related to https://github.com/joemasilotti/railsdevs.com/discussions/61

**Reply by joemasilotti** [reply]
Have you seen #820 and #868?
